src/directives/tooltip.js

```diff
@@ -0,0 +1,329 @@
+is asynchronous, so you must use `await`. For example:
+
+  await component.should.have.tooltip('Some text');
+*/
+
+import { computePosition, detectOverflow, flip, inline, offset } from '@floating-ui/dom';
+
+import { noop } from '../util/util';
+
+const types = ['text', 'aria-describedby', 'aria-label', 'sr-only', 'no-aria'];
+const placements = ['top', 'right', 'bottom', 'left'];
+const tooltips = new Map();
+let id = 1;
+
+// This Floating UI middleware is needed to horizontally center a v-tooltip.text
+// tooltip above or below its anchor element.
+const detectTextOverflow = {
+  name: 'detectTextOverflow',
+  fn: async (middlewareArgs) => {
+    const { placement } = middlewareArgs;
+    if (!(placement === 'top' || placement === 'bottom')) return {};
+    const overflow = await detectOverflow(middlewareArgs, {
+      elementContext: 'reference'
+    });
+    return {
+      // Accounting for both overflow.right and overflow.left with the thought
+      // that we will likely support RTL text in the future.
+      x: middlewareArgs.x -
+        (overflow.right > 0 ? overflow.right / 2 : 0) +
+        (overflow.left > 0 ? overflow.left / 2 : 0)
+    };
+  }
+};
+
+const show = (event) => {
+  const tooltip = tooltips.get(event.target);
+  if (!tooltip.shown) tooltip.show();
+};
+const hide = (event) => {
+  const tooltip = tooltips.get(event.target);
+  if (tooltip.shown) tooltip.hide();
+};
+
+class Tooltip {
+  static maybeCreate(anchor, { modifiers, value }) {
+    const type = types.find(name => modifiers[name] === true);
+    if (type == null) {
+      // eslint-disable-next-line no-console
+      console.error('invalid tooltip', anchor);
+      return null;
+    }
+
+    const tooltip = new Tooltip(anchor, type, value);
+    /*
+    For v-tooltip.text, we check whether the text is more than 5 characters
+    rather than checking whether tooltip.text() != null because the element that
+    sets text-overflow might not be visible yet. In that case, clientWidth === 0
+    on that element, so tooltip.text() == null even if the text will be
+    truncated once the element is visible. Rather than trying to watch the
+    visibility of the element, we add event listeners if the text is more than 5
+    characters. If/when an event listener is triggered for show, we will
+    evaluate then whether the text is truncated and show the tooltip if so.
+    Presumably the text won't ever be truncated if it is 5 characters or fewer.
+
+    For v-tooltip.sr-only, we always add event listeners because we can't easily
+    hook into the lifecycle of the .sr-only element. If/when an event listener
+    is triggered for show, we will evaluate then whether there is an .sr-only
+    element and whether it has text and show the tooltip if so.
+
+    For other tooltip types, we can know right away whether a tooltip should be
+    shown: we just need to check tooltip.text().
+    */
+    if (type === 'text'
+      ? anchor.textContent.trim().length > 5
```

Comment:
should we use innerText? textContent return content of the hidden children as well

Reply:
Ooh good question. As we use `v-tooltip` today, I can't think of cases where the two would differ. For example, I'm pretty sure we only use `v-tooltip` on elements without children. It usually works best to place `v-tooltip` on the deepest possible element so that the tooltip is positioned correctly. `.should.have.textTooltip()` even makes the assertion that the element has no children.

That said, I could imagine cases in the future where the two would be different and where it'd be better to use `innerText`. I'll go ahead and use `innerText` instead of `textContent` when computing the tooltip text in `_text()`.

On this particular line, I think I want to continue using `textContent`. MDN [notes](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#differences_from_innertext):

> since `innerText` takes CSS styles into account, reading the value of `innerText` triggers a reflow to ensure up-to-date computed styles. (Reflows can be computationally expensive, and thus should be avoided when possible.)

This line is part of the `Tooltip.maybeCreate()` method, which is called for every element that is a candidate to have a tooltip. In some cases, the number of candidates could be much greater than the number of elements to which we ultimately add a tooltip. I think we should make this candidate evaluation step as efficient as possible. On the other hand, `_text()` will only use `innerText` when a tooltip is being shown or when a tooltip that is currently shown is updated. Only one or two tooltips should be shown at a time, so there aren't the same performance considerations there.